extensions/smallrye-context-propagation/deployment/src/main/java/io/quarkus/smallrye/context/deployment/SmallRyeContextPropagationProcessor.java

```diff
@@ -59,12 +60,15 @@ void registerBean(BuildProducer<AdditionalBeanBuildItem> additionalBeans) {
 
     @BuildStep
     @Record(ExecutionTime.STATIC_INIT)
-    void buildStatic(SmallRyeContextPropagationRecorder recorder)
+    void buildStatic(SmallRyeContextPropagationRecorder recorder, List<ThreadContextProviderBuildItem> threadContextProviders)
             throws ClassNotFoundException, IOException {
         List<ThreadContextProvider> discoveredProviders = new ArrayList<>();
         List<ContextManagerExtension> discoveredExtensions = new ArrayList<>();
-        for (Class<?> provider : ServiceUtil.classesNamedIn(Thread.currentThread().getContextClassLoader(),
-                "META-INF/services/" + ThreadContextProvider.class.getName())) {
+        List<Class<?>> providers = threadContextProviders.stream().map(ThreadContextProviderBuildItem::getProvider)
+                .collect(Collectors.toList());
```

Comment:
While `Collectors.toList()` technically results in a simple `ArrayList`, its javadoc says "There are no guarantees on the type, mutability, serializability, or thread-safety of the `List` returned". Maybe this should be `.collect(Collectors.toCollection(ArrayList::new))`? (Yes, it is uglier. I don't insist.)

Reply:
Well, I would expect a LOT of compatibility issues if they change the behavior... BTW given the fact that we have `Collectors.toUnmodifiableList()` since jdk 10 it's kind of expected that the original method returns a mutable list. But ofc you're right, technically it's wrong.